README.md
<div align="center">

# Gifski Command

### Node.js module for [gifski](https://github.com/ImageOptim/gifski) GIF encoder CLI

[![NPM](https://nodei.co/npm/gifski-command.png?compact=true)](https://nodei.co/npm/gifski-command/)
<br />
[![](https://img.shields.io/npm/dt/gifski-command.svg?style=flat-square)](https://www.npmjs.com/package/gifski-command)

</div>

[![NPM Version](https://badgen.net/npm/v/gifski-command)](https://npmjs.org/package/gifski-command)
[![Coverage Status](https://coveralls.io/repos/github/pichillilorenzo/gifski-command/badge.svg?branch=main)](https://coveralls.io/github/pichillilorenzo/gifski-command?branch=main)
[![license](https://img.shields.io/github/license/pichillilorenzo/gifski-command)](/LICENSE)
[![Donate to this project using Paypal](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://www.paypal.me/LorenzoPichilli)

## Getting started

To get started with this library, you need to install it and add it to your project.

### Installation

```bash
# npm
npm install gifski-command --save

# yarn
yarn add gifski-command
```

## Library Usage

API Reference available at [https://pichillilorenzo.github.io/gifski-command/](https://pichillilorenzo.github.io/gifski-command/).

```javascript
import * as path from 'path';
import {GifskiCommand} from 'gifski-command';
// or
const path = require('path');
const {GifskiCommand} = require('gifski-command');

const command = new GifskiCommand({
  output: path.join(__dirname, 'test', 'video.gif'),
  frames: [ path.join(__dirname, 'test', 'video.mp4.frame*.png') ],
  quality: 100
});
command.on('progress', progress => {
  console.log(progress);
});
const result = await command.run();
```

## CLI Usage

- `<pattern>` represents a [glob pattern](https://www.npmjs.com/package/glob) used to specify PNG frames used by [gifski](https://github.com/ImageOptim/gifski). Surround the glob pattern with quotes (example `'./test/**/*.png'`).
- `[options]` are the same as the [gifski](https://github.com/ImageOptim/gifski) CLI options.

```
Usage: gifski-command [options] <pattern>

Examples: 
  gifski-command -Q 100 -o './test/video.gif' './test/**/video.mp4.frame*.png'
  gifski-command -o - './test/**/video.mp4.frame*.png' > './test/video.gif'

Arguments:
  pattern                Glob pattern of PNG image files. Surround the glob pattern with quotes (example './test/**/*.png').

Options:
  -o, --output <a.gif>   Destination file to write to; "-" means stdout
  -r, --fps <num>        Frame rate of animation. This means the speed, as all frames are kept. [default: 20]
  --fast                 50% faster encoding, but 10% worse quality and larger file size
  --extra                50% slower encoding, but 1% better quality
  -Q, --quality <1-100>  Lower quality may give smaller file [default: 90]
  -W, --width <px>       Maximum width. By default anims are limited to about 800x600
  -H, --height <px>      Maximum height (stretches if the width is also set)
  --no-sort              Use files exactly in the order given, rather than sorted
  -q, --quiet            Do not display anything on standard output/console
  --repeat <num>         Number of times the animation is repeated (-1 none, 0 forever or <value> repetitions)
  -h, --help             display help for command
```

## CLI Usage Example

The recommended way is to first export video as PNG frames. If you have [ffmpeg](https://ffmpeg.org/) installed, you can run in terminal:

```bash
ffmpeg -i ./test/video.mp4 ./test/video.mp4.frame%04d.png
```

and then make the GIF from the frames:

```bash
gifski-command -Q 100 -o './test/video.gif' './test/**/video.mp4.frame*.png'
```

This code snippet shows how to put into action `gifski-command` to convert video frames into a high quality GIF using the [gifski](https://github.com/ImageOptim/gifski) encoder.

## Contributors

Any contribution is appreciated. You can get started with the steps below:

1. Fork [this repository](https://github.com/pichillilorenzo/gifski-command) (learn how to do this [here](https://help.github.com/articles/fork-a-repo)).

2. Clone the forked repository.

3. Make your changes and create a pull request ([learn how to do this](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)).

4. I will attend to your pull request and provide some feedback.

## License

This repository is licensed under the [ISC](LICENSE) License.


CHANGELOG.md
## 1.0.6

- Added glob pattern support to `GifskiCommandOptions.frames` property.

## 1.0.5

- Updated `tsconfig.json` to maintain comments.

## 1.0.4

- Fixed `gifskiPath`

## 1.0.3

- Fixed `gifskiPath`

## 1.0.2

- Updated `GifskiCommandOptions.quality` type to `number`

## 1.0.1

- Updated README
- Fixed `--quiet` CLI flag
- Added `runSync` method

## 1.0.0

Initial release


test/lib.test.ts
import {describe, expect, test} from '@jest/globals';
import {GifskiCommand} from "../src";
import * as path from "path";
import * as fs from "fs";
import {spawnSync} from "child_process";

const output = path.join(__dirname, 'video.gif');
const output2 = path.join(__dirname, 'video2.gif');
const frames = [path.join(__dirname, 'video.mp4.frame*.png')];

jest.setTimeout(20000);

describe('gifski-command', () => {
  beforeEach(cb => {
    if (fs.existsSync(output)) {
      fs.unlinkSync(output);
    }
    if (fs.existsSync(output2)) {
      fs.unlinkSync(output2);
    }
    cb();
  });

  afterEach(cb => {
    if (fs.existsSync(output)) {
      fs.unlinkSync(output);
    }
    if (fs.existsSync(output2)) {
      fs.unlinkSync(output2);
    }
    cb();
  });

  test('basic usage', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
  });

  test('basic usage with glob pattern frames', async () => {
    const command = new GifskiCommand({
      output,
      frames: path.join(__dirname, '**', 'video.mp4.frame*.png')
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
  });

  test('basic usage error', async () => {
    const command = new GifskiCommand({
      output,
      frames: [path.join(__dirname, 'video.mp4.frame*.not.existing')]
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeFalsy();
    expect(result.err).toBeDefined();
    expect(result.stderr).toBeDefined();
  });

  test('basic usage sync', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    const result = command.runSync();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
  });

  test('basic usage to stdout', async () => {
    const command = new GifskiCommand({
      output: '-',
      frames
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeFalsy();
    expect(result.err).toBeUndefined();
    expect(result.stdout).toBeDefined();
  });

  test('basic usage sync to stdout', async () => {
    const command = new GifskiCommand({
      output: '-',
      frames
    });
    const result = command.runSync();
    expect(fs.existsSync(output)).toBeFalsy();
    expect(result.err).toBeUndefined();
    expect(result.stdout).toBeDefined();
  });

  test('progress event', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    command.on('progress', progress => {
      expect(progress).toBeDefined();
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
  });

  test('progress event not called with sync', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    let i = 0;
    command.on('progress', () => {
      i++;
    });
    const result = command.runSync();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
    expect(i).toBe(0);
  });

  test('end event', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    command.on('end', (err) => {
      expect(err).toBeUndefined();
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
  });

  test('end event not called with sync', async () => {
    const command = new GifskiCommand({
      output,
      frames
    });
    let i = 0;
    command.on('end', () => {
      i++;
    });
    const result = command.runSync();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
    expect(i).toBe(0);
  });

  test('quality', async () => {
    const command = new GifskiCommand({
      output,
      frames,
      quality: 100
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
    const {size} = fs.statSync(output);

    const command2 = new GifskiCommand({
      output: output2,
      frames,
      quality: 10
    });
    const result2 = await command2.run();
    expect(fs.existsSync(output2)).toBeTruthy();
    expect(result2.err).toBeUndefined();
    const {size: size2} = fs.statSync(output2);

    expect(size).toBeGreaterThan(size2);
  });

  test('quiet', async () => {
    const command = new GifskiCommand({
      output,
      frames,
      quiet: true
    });
    const result = await command.run();
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.err).toBeUndefined();
    expect(result.stdout).toBeUndefined();
  });
});

describe('gifski-command CLI', () => {
  beforeEach(cb => {
    if (fs.existsSync(output)) {
      fs.unlinkSync(output);
    }
    if (fs.existsSync(output2)) {
      fs.unlinkSync(output2);
    }
    cb();
  });

  afterEach(cb => {
    if (fs.existsSync(output)) {
      fs.unlinkSync(output);
    }
    if (fs.existsSync(output2)) {
      fs.unlinkSync(output2);
    }
    cb();
  });

  test('basic usage', async () => {
    const result = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '-o', output,
        ...frames
      ], {
        stdio: 'inherit'
      }
    );
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.error).toBeUndefined();
  });

  test('basic usage to stdout', async () => {
    const result = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '-o', '-',
        ...frames
      ], {
        stdio: 'inherit'
      }
    );
    expect(fs.existsSync(output)).toBeFalsy();
    expect(result.error).toBeUndefined();
    expect(result.stdout).toBeDefined();
  });

  test('quality', async () => {
    const result = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '-Q', '100',
        '-o', output,
        ...frames
      ], {
        stdio: 'inherit'
      }
    );
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.error).toBeUndefined();
    const {size} = fs.statSync(output);

    const result2 = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '-Q', '10',
        '-o', output2,
        ...frames
      ], {
        stdio: 'inherit'
      }
    );
    expect(fs.existsSync(output2)).toBeTruthy();
    expect(result2.error).toBeUndefined();
    const {size: size2} = fs.statSync(output2);

    expect(size).toBeGreaterThan(size2);
  });

  test('quiet', async () => {
    const result = spawnSync('ts-node', [
        path.join(__dirname, '..', 'src', 'cli.ts'),
        '--quiet',
        '-o', output,
        ...frames
      ], {
        stdio: 'inherit'
      }
    );
    expect(fs.existsSync(output)).toBeTruthy();
    expect(result.error).toBeUndefined();
    expect(result.stdout).toBeNull();
  });
});


lib/cli.js
#!/usr/bin/env node
"use strict";Object.defineProperty(exports,"__esModule",{value:!0});const tslib_1=require("tslib"),commander_1=require("commander"),glob_1=require("glob"),index_1=require("./index"),path=tslib_1.__importStar(require("path")),program=new commander_1.Command,Bar=require("progress-barjs");function log(e,...t){e.output!=="-"&&!e.quiet&&console.log(...t)}async function generateGIF(e){const t={...program.opts()};if(e&&t.output){const o=glob_1.glob.sync(`${e}`,{absolute:!0});t.output!=="-"&&(t.output=path.resolve(t.output)),log(t,`${o.length} PNG image file${o.length!==1?"s":""} found!`);const i=new index_1.GifskiCommand({...t,frames:o});if(t.output!=="-"&&!t.quiet){const u=Bar({label:"Convert to .gif",info:"Processing",total:o.length});let a=0;i.on("progress",s=>{u.tickChunk(s.currentFrame-a),a=s.currentFrame})}const{err:r,stdout:n}=await i.run();if(r){console.error(r),process.exit(1);return}log(t,`\u{1F389} ${o.length} PNG image file${o.length!==1?"s have":"had"} been converted with success. \u{1F389}`),log(t,`GIF saved at: ${t.output}`),t.output==="-"&&n instanceof Buffer&&process.stdout.write(n)}else{if(!e)throw Error("Empty glob pattern of PNG image files!");if(!t.output)throw Error("Empty output destination file path!")}}async function run(){program.argument("<pattern>","Glob pattern of PNG image files. Surround the glob pattern with quotes (example './test/**/*.png').").description(`Examples: 
  gifski-command -Q 100 -o './test/video.gif' './test/**/video.mp4.frame*.png'
  gifski-command -o - './test/**/video.mp4.frame*.png' > './test/video.gif'`).requiredOption("-o, --output <a.gif>",'Destination file to write to; "-" means stdout').option("-r, --fps <num>","Frame rate of animation. This means the speed, as all frames are kept. [default: 20]").option("--fast","50% faster encoding, but 10% worse quality and larger file size").option("--extra","50% slower encoding, but 1% better quality").option("-Q, --quality <1-100>","Lower quality may give smaller file [default: 90]").option("-W, --width <px>","Maximum width. By default anims are limited to about 800x600").option("-H, --height <px>","Maximum height (stretches if the width is also set)").option("--no-sort","Use files exactly in the order given, rather than sorted").option("-q, --quiet","Do not display anything on standard output/console").option("--repeat <num>","Number of times the animation is repeated (-1 none, 0 forever or <value> repetitions)").action(generateGIF),await program.parseAsync()}run();

lib/index.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0}),exports.GifskiCommand=exports.gifskiPath=void 0;const tslib_1=require("tslib"),child_process_1=require("child_process"),os=tslib_1.__importStar(require("os")),path=tslib_1.__importStar(require("path")),events_1=tslib_1.__importDefault(require("events")),fs=tslib_1.__importStar(require("fs")),glob_1=require("glob"),platform=process.env.npm_config_platform||os.platform();let platformPath="";switch(platform){case"win32":platformPath=path.join("windows","gifski.exe");break;case"darwin":platformPath=path.join("macos","gifski");break;case"linux":platformPath=path.join("linux","gifski");break}if(!platformPath)throw Error(`gifski executable not found for platform ${platform}.`);let _gifskiPath=path.join(__dirname,"..","node_modules","gifski","bin",platformPath);if(fs.existsSync(_gifskiPath)||(_gifskiPath=path.join(__dirname,"..","..","..","node_modules","gifski","bin",platformPath)),!fs.existsSync(_gifskiPath))throw Error(`gifski executable not found in ${_gifskiPath}.`);exports.gifskiPath=_gifskiPath;class GifskiCommand extends events_1.default.EventEmitter{options;constructor(t){super(),this.options=t}_buildSpawnArgs(){const t=[];this.options.fps&&t.push("--fps",this.options.fps.toString()),this.options.fast&&t.push("--fast"),this.options.extra&&t.push("--extra"),this.options.quality&&t.push("--quality",this.options.quality.toString()),this.options.width&&t.push("--width",this.options.width.toString()),this.options.height&&t.push("--height",this.options.height.toString()),this.options.noSort&&t.push("--no-sort"),this.options.quiet&&t.push("--quiet"),this.options.repeat&&t.push("--repeat",this.options.repeat.toString());let s=this.options.frames;return typeof s=="string"&&(s=glob_1.glob.sync(s,{absolute:!0})),t.push("-o",this.options.output,...s),t}async run(t){t={shell:!0,...t};const s=this.options.stdoutLines??100,i=[];let u=!1;const n=[];let l=!1,d=!1,h;return new Promise((f,_)=>{const p=o=>{if(o&&(h=o),d&&u&&l){const e=i.length>0?this.options.output!=="-"?i.join(`
`):Buffer.concat(i):void 0,r=n.length>0?n.join(`
`):void 0;h&&h.message.match(/gifski exited with code/)&&r&&(h.message+=": "+r),this.emit("end",h,e,r),f({err:h,stdout:e,stderr:r})}};let a=(0,child_process_1.spawn)(`${exports.gifskiPath}`,this._buildSpawnArgs(),t);a.stdout.on("data",o=>{this.options.output!=="-"?this.options.quiet||(s!==0&&i.length===s&&i.shift(),i.push(o.toString())):i.push(o);const e=[...o.toString().trim().matchAll(/^Frame (\d+) \/ (\d+).*(\d*)s*$/g)][0];if(e){const r=parseInt(e[1]),g=parseInt(e[2]),c=e[3]?parseInt(e[3]):0,m={currentFrame:r,totalFrames:g,percent:r/g*100,secondsLeft:c};this.emit("progress",m)}}),a.stdout.on("close",()=>{u=!0,p()}),a.stderr.on("data",o=>{s!==0&&n.length===s&&n.shift(),n.push(o.toString())}),a.stderr.on("close",()=>{l=!0,p()}),a.on("error",o=>{const e=i.length>0?this.options.output!=="-"?i.join(`
`):Buffer.concat(i):void 0,r=n.length>0?n.join(`
`):void 0;this.emit("error",o,e,r)}),a.on("exit",(o,e)=>{d=!0,e?p(new Error("gifski was killed with signal "+e)):o?p(new Error("gifski exited with code "+o)):p(),a=null})})}runSync(t){t={shell:!0,stdio:"inherit",...t};const s=(0,child_process_1.spawnSync)(`${exports.gifskiPath}`,this._buildSpawnArgs(),t);let i=s.error;return s.signal&&s.signal!=="SIGTERM"?i?i.message+=`: gifski was killed with signal ${s.signal}`:i=new Error(`gifski was killed with signal ${s.signal}`):s.status&&(i?i.message+=`: gifski exited with code ${s.status}`:i=new Error(`gifski exited with code ${s.status}`)),{err:i,stdout:s.stdout,stderr:s.stderr}}}exports.GifskiCommand=GifskiCommand;

src/index.ts
import {
  ChildProcessWithoutNullStreams,
  spawn,
  SpawnOptionsWithoutStdio,
  spawnSync,
  SpawnSyncOptionsWithBufferEncoding, SpawnSyncReturns
} from "child_process";
import * as os from "os";
import * as path from "path";
import events from "events";
import * as fs from "fs";
import {glob} from "glob";

const platform = process.env["npm_config_platform"] || os.platform();

let platformPath = '';
switch (platform) {
  case 'win32':
    platformPath = path.join('windows', 'gifski.exe');
    break;
  case 'darwin':
    platformPath = path.join('macos', 'gifski');
    break;
  case 'linux':
    platformPath = path.join('linux', 'gifski');
    break;
}
if (!platformPath) {
  throw Error(`gifski executable not found for platform ${platform}.`);
}

let _gifskiPath = path.join(
  __dirname, '..', 'node_modules', 'gifski', 'bin', platformPath
);
if (!fs.existsSync(_gifskiPath)) {
  _gifskiPath = path.join(
    __dirname, '..', '..', '..', 'node_modules', 'gifski', 'bin', platformPath
  );
}
if (!fs.existsSync(_gifskiPath)) {
  throw Error(`gifski executable not found in ${_gifskiPath}.`);
}

/**
 * gifski executable path.
 */
export const gifskiPath = _gifskiPath;

/**
 * gifski command options.
 */
export interface GifskiCommandOptions {
  /**
   * <a.gif>. Destination file to write to; "-" means stdout.
   */
  output: string;

  /**
   * Glob pattern or list of PNG image files.
   */
  frames: string | string[];

  /**
   * Frame rate of animation. This means the speed, as all frames are kept. [default: 20].
   */
  fps?: number;

  /**
   * 50% faster encoding, but 10% worse quality and larger file size.
   */
  fast?: boolean;

  /**
   * 50% slower encoding, but 1% better quality.
   */
  extra?: boolean;

  /**
   * <1-100>. Lower quality may give smaller file [default: 90].
   */
  quality?: number;

  /**
   * <px>. Maximum width. By default anims are limited to about 800x600.
   */
  width?: number;

  /**
   * <px>. Maximum height (stretches if the width is also set).
   */
  height?: number;

  /**
   * Use files exactly in the order given, rather than sorted.
   */
  noSort?: boolean;

  /**
   * Do not display anything on standard output/console
   */
  quiet?: boolean;

  /**
   * <num> Number of times the animation is repeated (-1 none, 0 forever or <value> repetitions).
   */
  repeat?: number;

  /**
   * Maximum number of lines from gifski stdout/stderr to keep in memory (defaults to 100, use 0 for unlimited storage).
   */
  stdoutLines?: number;
}

/**
 * gifski progress event interface.
 */
export interface GifskiCommandProgress {
  /**
   * Current Frame number being processed.
   */
  currentFrame: number;

  /**
   * Total number of frames.
   */
  totalFrames: number;

  /**
   * Progress percentage.
   */
  percent: number;

  /**
   * Seconds left to completion.
   */
  secondsLeft: number;
}

export declare interface GifskiCommand {
  /**
   * Event called multiple times when gifski is running.
   *
   * @param event
   * @param listener
   */
  on(event: 'progress', listener: (name: GifskiCommandProgress) => void): this;

  /**
   * Event called when gifski command is completed.
   *
   * @param event
   * @param listener
   */
  on(event: 'end', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;

  /**
   * Event called when gifski command emit an error.
   *
   * @param event
   * @param listener
   */
  on(event: 'error', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;
}

/**
 * gifski command wrapper.
 */
export class GifskiCommand extends events.EventEmitter {
  /**
   * Options used by the gifski command.
   */
  options: GifskiCommandOptions;

  constructor(options: GifskiCommandOptions) {
    super();
    this.options = options;
  }

  private _buildSpawnArgs(): string[] {
    const processArgs: string[] = [];

    if (this.options.fps) {
      processArgs.push(
        '--fps',
        this.options.fps.toString()
      );
    }
    if (this.options.fast) {
      processArgs.push('--fast');
    }
    if (this.options.extra) {
      processArgs.push('--extra');
    }
    if (this.options.quality) {
      processArgs.push(
        '--quality',
        this.options.quality.toString()
      );
    }
    if (this.options.width) {
      processArgs.push(
        '--width',
        this.options.width.toString()
      );
    }
    if (this.options.height) {
      processArgs.push(
        '--height',
        this.options.height.toString()
      );
    }
    if (this.options.noSort) {
      processArgs.push('--no-sort');
    }
    if (this.options.quiet) {
      processArgs.push('--quiet');
    }
    if (this.options.repeat) {
      processArgs.push(
        '--repeat',
        this.options.repeat.toString()
      );
    }

    let frames = this.options.frames;
    if (typeof frames === 'string') {
      frames = glob.sync(frames, {
        absolute: true
      });
    }

    processArgs.push(
      '-o',
      this.options.output,
      ...frames
    );
    return processArgs;
  }

  /**
   * Run gifski command.
   *
   * @param options
   */
  async run(options?: SpawnOptionsWithoutStdio): Promise<{ err?: Error, stdout?: string | Buffer, stderr?: string | Buffer }> {
    options = {
      shell: true,
      ...options
    };

    const maxLines = this.options.stdoutLines ?? 100;

    const stdoutData: (string | Buffer)[] = [];
    let stdoutClosed = false;

    const stderrData: string[] = [];
    let stderrClosed = false;

    let processExited = false;
    let exitError: Error | undefined;

    return new Promise((resolve, _) => {
      const _handleExit = (err?: Error) => {
        if (err) {
          exitError = err;
        }

        if (processExited && stdoutClosed && stderrClosed) {
          const stdout = stdoutData.length > 0 ?
            (this.options.output !== '-' ? stdoutData.join('\n') : Buffer.concat(stdoutData as Buffer[])) :
            undefined;
          const stderr = stderrData.length > 0 ? stderrData.join('\n') : undefined;
          if (exitError && exitError.message.match(/gifski exited with code/) && stderr) {
            // Add gifski error message
            exitError.message += ': ' + stderr;
          }
          this.emit('end', exitError, stdout, stderr);
          resolve({err: exitError, stdout, stderr});
        }
      }

      let gifskiProcess: ChildProcessWithoutNullStreams | null = spawn(
        `${gifskiPath}`,
        this._buildSpawnArgs(),
        options
      );

      gifskiProcess.stdout.on('data', chunk => {
        if (this.options.output !== '-') {
          if (!this.options.quiet) {
            if (maxLines !== 0 && stdoutData.length === maxLines) {
              stdoutData.shift();
            }
            stdoutData.push(chunk.toString());
          }
        } else {
          stdoutData.push(chunk);
        }

        const matches = [...chunk.toString().trim().matchAll(/^Frame (\d+) \/ (\d+).*(\d*)s*$/g)][0];
        if (matches) {
          const currentFrame = parseInt(matches[1]);
          const totalFrames = parseInt(matches[2]);
          const secondsLeft = matches[3] ? parseInt(matches[3]) : 0;

          const progress: GifskiCommandProgress = {
            currentFrame,
            totalFrames,
            percent: currentFrame / totalFrames * 100,
            secondsLeft
          };

          this.emit('progress', progress);
        }
      });

      gifskiProcess.stdout.on('close', () => {
        stdoutClosed = true;
        _handleExit();
      });

      gifskiProcess.stderr.on('data', chunk => {
        if (maxLines !== 0 && stderrData.length === maxLines) {
          stderrData.shift();
        }
        stderrData.push(chunk.toString());
      });

      gifskiProcess.stderr.on('close', () => {
        stderrClosed = true;
        _handleExit();
      });

      gifskiProcess.on('error', err => {
        const stdout = stdoutData.length > 0 ?
          (this.options.output !== '-' ? stdoutData.join('\n') : Buffer.concat(stdoutData as Buffer[])) :
          undefined;
        const stderr = stderrData.length > 0 ? stderrData.join('\n') : undefined;

        this.emit('error', err, stdout, stderr);
      });

      gifskiProcess.on('exit', (code, signal) => {
        processExited = true;

        if (signal) {
          _handleExit(new Error('gifski was killed with signal ' + signal));
        } else if (code) {
          _handleExit(new Error('gifski exited with code ' + code));
        } else {
          _handleExit();
        }

        gifskiProcess = null;
      });
    });
  }

  /**
   * Run gifski command in sync way.
   *
   * @param options
   */
  runSync(options?: SpawnSyncOptionsWithBufferEncoding): { err?: Error, stdout?: string | Buffer, stderr?: string | Buffer } {
    options = {
      shell: true,
      stdio: 'inherit',
      ...options
    };

    const result: SpawnSyncReturns<string | Buffer> = spawnSync(
      `${gifskiPath}`,
      this._buildSpawnArgs(),
      options
    );

    let err = result.error;
    if (result.signal && result.signal !== 'SIGTERM') {
      if (err) {
        err.message += `: gifski was killed with signal ${result.signal}`;
      } else {
        err = new Error(`gifski was killed with signal ${result.signal}`);
      }
    } else if (result.status) {
      if (err) {
        err.message += `: gifski exited with code ${result.status}`;
      } else {
        err = new Error(`gifski exited with code ${result.status}`);
      }
    }

    return {
      err,
      stdout: result.stdout,
      stderr: result.stderr
    };
  }
}


lib/index.d.ts
/// <reference types="node" />
/// <reference types="node" />
/// <reference types="node" />
import { SpawnOptionsWithoutStdio, SpawnSyncOptionsWithBufferEncoding } from "child_process";
import events from "events";
/**
 * gifski executable path.
 */
export declare const gifskiPath: string;
/**
 * gifski command options.
 */
export interface GifskiCommandOptions {
    /**
     * <a.gif>. Destination file to write to; "-" means stdout.
     */
    output: string;
    /**
     * Glob pattern or list of PNG image files.
     */
    frames: string | string[];
    /**
     * Frame rate of animation. This means the speed, as all frames are kept. [default: 20].
     */
    fps?: number;
    /**
     * 50% faster encoding, but 10% worse quality and larger file size.
     */
    fast?: boolean;
    /**
     * 50% slower encoding, but 1% better quality.
     */
    extra?: boolean;
    /**
     * <1-100>. Lower quality may give smaller file [default: 90].
     */
    quality?: number;
    /**
     * <px>. Maximum width. By default anims are limited to about 800x600.
     */
    width?: number;
    /**
     * <px>. Maximum height (stretches if the width is also set).
     */
    height?: number;
    /**
     * Use files exactly in the order given, rather than sorted.
     */
    noSort?: boolean;
    /**
     * Do not display anything on standard output/console
     */
    quiet?: boolean;
    /**
     * <num> Number of times the animation is repeated (-1 none, 0 forever or <value> repetitions).
     */
    repeat?: number;
    /**
     * Maximum number of lines from gifski stdout/stderr to keep in memory (defaults to 100, use 0 for unlimited storage).
     */
    stdoutLines?: number;
}
/**
 * gifski progress event interface.
 */
export interface GifskiCommandProgress {
    /**
     * Current Frame number being processed.
     */
    currentFrame: number;
    /**
     * Total number of frames.
     */
    totalFrames: number;
    /**
     * Progress percentage.
     */
    percent: number;
    /**
     * Seconds left to completion.
     */
    secondsLeft: number;
}
export declare interface GifskiCommand {
    /**
     * Event called multiple times when gifski is running.
     *
     * @param event
     * @param listener
     */
    on(event: 'progress', listener: (name: GifskiCommandProgress) => void): this;
    /**
     * Event called when gifski command is completed.
     *
     * @param event
     * @param listener
     */
    on(event: 'end', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;
    /**
     * Event called when gifski command emit an error.
     *
     * @param event
     * @param listener
     */
    on(event: 'error', listener: (err?: Error, stdout?: string | Buffer, stderr?: string | Buffer) => void): this;
}
/**
 * gifski command wrapper.
 */
export declare class GifskiCommand extends events.EventEmitter {
    /**
     * Options used by the gifski command.
     */
    options: GifskiCommandOptions;
    constructor(options: GifskiCommandOptions);
    private _buildSpawnArgs;
    /**
     * Run gifski command.
     *
     * @param options
     */
    run(options?: SpawnOptionsWithoutStdio): Promise<{
        err?: Error;
        stdout?: string | Buffer;
        stderr?: string | Buffer;
    }>;
    /**
     * Run gifski command in sync way.
     *
     * @param options
     */
    runSync(options?: SpawnSyncOptionsWithBufferEncoding): {
        err?: Error;
        stdout?: string | Buffer;
        stderr?: string | Buffer;
    };
}


src/cli.ts
#!/usr/bin/env node
import {Command} from 'commander';
import {glob} from 'glob';
import {GifskiCommand, GifskiCommandOptions} from "./index";
import * as path from "path";

const program = new Command();

const Bar = require('progress-barjs');

function log(options: GifskiCommandOptions, ...args: any[]) {
  if (options.output !== '-' && !options.quiet) {
    console.log(...args);
  }
}

async function generateGIF(pattern: string) {
  const options: GifskiCommandOptions = {
    ...program.opts()
  };
  if (pattern && options.output) {
    const files = glob.sync(`${pattern}`, {
      absolute: true
    });

    if (options.output !== '-') {
      options.output = path.resolve(options.output);
    }

    log(options, `${files.length} PNG image file${files.length !== 1 ? 's' : ''} found!`);

    const command = new GifskiCommand({
      ...options,
      frames: files
    });

    if (options.output !== '-' && !options.quiet) {
      const bar = Bar({
        label: `Convert to .gif`,
        info: 'Processing',
        total: files.length,
      });
      let previousCurrentFrame = 0;
      command.on('progress', progress => {
        bar.tickChunk(progress.currentFrame - previousCurrentFrame);
        previousCurrentFrame = progress.currentFrame;
      });
    }

    const {err, stdout} = await command.run();
    if (err) {
      console.error(err);
      process.exit(1);
      return;
    }

    log(options, `🎉 ${files.length} PNG image file${files.length !== 1 ? 's have' : 'had'} been converted with success. 🎉`);
    log(options, `GIF saved at: ${options.output}`);

    if (options.output === '-' && stdout instanceof Buffer) {
      process.stdout.write(stdout as Buffer);
    }
  } else {
    if (!pattern)
      throw Error('Empty glob pattern of PNG image files!');
    if (!options.output)
      throw Error('Empty output destination file path!');
  }
}

async function run() {
  program
    .argument('<pattern>', `Glob pattern of PNG image files. Surround the glob pattern with quotes (example './test/**/*.png').`)
    .description(`Examples: 
  gifski-command -Q 100 -o './test/video.gif' './test/**/video.mp4.frame*.png'
  gifski-command -o - './test/**/video.mp4.frame*.png' > './test/video.gif'`)
    .requiredOption('-o, --output <a.gif>', 'Destination file to write to; "-" means stdout')
    .option('-r, --fps <num>', 'Frame rate of animation. This means the speed, as all frames are kept. [default: 20]')
    .option('--fast', '50% faster encoding, but 10% worse quality and larger file size')
    .option('--extra', '50% slower encoding, but 1% better quality')
    .option('-Q, --quality <1-100>', 'Lower quality may give smaller file [default: 90]')
    .option('-W, --width <px>', 'Maximum width. By default anims are limited to about 800x600')
    .option('-H, --height <px>', 'Maximum height (stretches if the width is also set)')
    .option('--no-sort', 'Use files exactly in the order given, rather than sorted')
    .option('-q, --quiet', 'Do not display anything on standard output/console')
    .option('--repeat <num>', 'Number of times the animation is repeated (-1 none, 0 forever or <value> repetitions)')
    .action(generateGIF);
  await program.parseAsync();
}

run();
